### PR TITLE
fix(core): do not write View clearColor into the picking buffer

### LINKED ISSUE #10430: [Bug] View clearColor corrupts the picking buffer, breaking object picking
### Description

Since the `View` `clear`/`clearColor` props were introduced in #9650 (v9.2), `LayersPass._drawLayersInViewport` applies the view's clear color to every render target, including the picking framebuffer used by `PickLayersPass`.

The picking buffer encodes picked object references as colors (the alpha channel encodes the layer index, see `getClosestObject` in `query-object.ts`). Clearing it with any color whose alpha is non-zero makes every background pixel decode as a "picked object" of a non-existent layer. Two user-visible consequences:

1. Picking with a `pickingRadius` fails whenever the pointer is not exactly on the object: the background pixel under the pointer (distance 0) wins over the real object pixel nearby, then fails to decode.
2. Every hover/click over the background logs `deck: Picked non-existent layer. Is picking buffer corrupt?`.

### Flavors

- [ ] Script tag
- [ ] React
- [ ] Python/Jupyter notebook
- [ ] MapboxOverlay
- [ ] GoogleMapsOverlay
- [ ] CARTO
- [ ] ArcGIS

### Expected Behavior

`View` `clearColor` should only affect the visible canvas. Picking should behave exactly as it does without `clearColor`: clicks within `pickingRadius` of an object pick it, background picks return nothing, and no errors are logged.

### Steps to Reproduce

```js
import {Deck, MapView} from '@deck.gl/core';
import {ScatterplotLayer} from '@deck.gl/layers';

new Deck({
  views: new MapView({clear: true, clearColor: [255, 255, 255, 255]}),
  initialViewState: {longitude: 0, latitude: 0, zoom: 1},
  pickingRadius: 20,
  layers: [
    new ScatterplotLayer({
      data: [{position: [0, 0]}],
      getPosition: d => d.position,
      radiusMinPixels: 10,
      getFillColor: [255, 0, 0],
      pickable: true
    })
  ],
  onClick: info => console.log('picked:', info.object ?? null)
});
```

1. Run the snippet with `deck.gl@9.3.6` and open the browser console.
2. Click just next to the red circle (within the 20px picking radius, but not on the circle).
3. Click anywhere on the white background.

**Observed:**

- Step 2 logs `picked: null` plus `deck: Picked non-existent layer. Is picking buffer corrupt?`, even though the click is within `pickingRadius` of the object.
- Step 3 also logs the `Picked non-existent layer` error on every click (and on every hover).

**Expected:** (matches the behavior after removing `clear: true, clearColor` from the `MapView`)

- Step 2 logs `picked: {position: [0, 0]}`.
- Step 3 logs `picked: null` with no errors.

#### Environment

- **Framework version**: deck.gl@9.3.6
- **Browser**: Chrome 150.0.7871.46
- **OS**: MacOS 26.5.2

### PR DESCRIPTION
**Pull request by jarnoharno**
Closes #10430

#### Background

Since #9650, `LayersPass._drawLayersInViewport` applies the `View` `clear`/`clearColor` props to every render target, including the picking framebuffer used by `PickLayersPass`. The picking buffer encodes picked object references as colors (alpha encodes the layer index), so clearing it with any color whose alpha is non-zero makes every background pixel decode as a picked object of a non-existent layer. As a result, picking within `pickingRadius` fails whenever the pointer is not exactly on an object, and every hover/click over the background logs `Picked non-existent layer. Is picking buffer corrupt?`.

#### Change List

- When picking, ignore the view's clearColor and let the clear fall through to transparent black, so background pixels keep decoding as "no object".  `clearColor: false` is still honored while picking.
- Add two tests:
  - A regression test that renders the picking buffer with a `MapView` using `clearColor: [255, 255, 255, 255]` and asserts that background pixels decode to "no object" while the object is still picked. Fails on master.
  - A guard for the `clearColor: false` boundary: two overlapping views where the top view uses `clear: true` + `clearColor: false`, asserting the bottom view's picked object survives. This is not a regression relative to master; it prevents the picking exception from being over-applied.

---

> [!NOTE]
> **Medium Risk**
> Touches shared layer rendering/clear logic used by picking; behavior change is scoped to picking mode with tests, but incorrect clear handling could still affect multi-view picking.
> 
> **Overview**
> Fixes picking corruption when a `View` uses `clear` with a non-transparent **`clearColor`**. View clears were being applied to the picking framebuffer, so background pixels no longer decoded as “no object” and hover/click within **`pickingRadius`** could fail.
> 
> **`LayersPass._drawLayersInViewport`** now receives **`isPicking`** and, while picking, **ignores the view’s `clearColor`** so the buffer stays cleared to transparent black; **`clearColor: false`** is still respected. Two **`PickLayersPass`** tests cover the white-clear regression and overlapping views with **`clearColor: false`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c845eb028abbac041966a5173ed735a8d64c05d0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: modules/core/src/passes/layers-pass.ts [hunk 1]
```diff
@@ -276,7 +279,12 @@ export default class LayersPass extends Pass {
...
-        if (Array.isArray(clearColor)) {
+        if (isPicking) {
+          // The picking buffer encodes object references as colors. Clearing with the
+          // view's clearColor would fill the background with a color that decodes to a
+          // (non-existent) picked object, so clear with transparent black instead.
+          colorToUse = [0, 0, 0, 0];
+        } else if (Array.isArray(clearColor)) {
           colorToUse = [...clearColor.slice(0, 3), clearColor[3] || 255].map(
             c => c / 255
           ) as NumberArray4;
```
Comment (chrisgervang) [inline modules/core/src/passes/layers-pass.ts]: Small refinement to preserve `clearColor` semantics during picking: rather than unconditionally clearing to transparent black, only substitute it when `clearColor` is an actual color array. That way a view with `clear: true` + `clearColor: false` still means "clear depth/stencil but don't touch color" while picking — the picking FBO is already cleared to transparent black at the top level (`pick-layers-pass.ts`), so the background still decodes as "no object" (alpha 0), and any overlapping-view picking data underneath is preserved, as `clearColor: false` promises.

Verified locally (headless Chromium): the new regression test still passes, and a `clear: true` + `clearColor: false` case also stays uncorrupted.

```suggestion
        if (Array.isArray(clearColor)) {
          // The picking buffer encodes object references in the color's alpha channel.
          // Clearing it with the view's clearColor would make the background decode as a
          // (non-existent) picked object, so clear with transparent black while picking.
          // Still honor `clearColor: false` (below) so "don't clear color" is preserved.
          colorToUse = isPicking
            ? [0, 0, 0, 0]
            : ([...clearColor.slice(0, 3), clearColor[3] || 255].map(
                c => c / 255
              ) as NumberArray4);
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/modules/core/passes/pick-layers-pass.spec.ts [hunk 2]
```diff
@@ -48,3 +48,55 @@ test('PickLayersPass#drawPickingBuffer', () => {
...
+  expect(
+    cornerColor[3],
+    'background pixel decodes to "no object" (alpha 0) despite view clearColor'
+  ).toBe(0);
+
+  const centerIndex = (50 * 100 + 50) * 4;
+  const centerColor = pixels.slice(centerIndex, centerIndex + 4);
+  const pickedObject = decodePickingColor!(centerColor);
+  expect(pickedObject?.pickedLayer?.id, 'object is still picked at the center').toBe(layer.id);
+});
```
Comment (chrisgervang) [inline test/modules/core/passes/pick-layers-pass.spec.ts:102]: If you take the `clearColor: false` refinement in the other comment, here's a regression test that actually guards it. It uses two overlapping views where the top view has `clear: true` + `clearColor: false`, and asserts the bottom view's picked object survives at a pixel inside the top view's rect. I verified this fails against the current `isPicking → [0,0,0,0]` version (the object gets wiped, decodes to `undefined`) and passes with the refinement — so it's a true guard, not one that passes either way.

```suggestion
});

test('PickLayersPass#overlapping view clearColor:false preserves underlying picking', () => {
  const pickingFBO = device.createFramebuffer({colorAttachments: ['rgba8unorm']});
  pickingFBO.resize({width: 100, height: 100});

  // Bottom view: full screen, no clear. Its object renders at screen (50, 50).
  const viewA = new MapView({id: 'A'});
  // Top view: right region x:40..100 with clear enabled but clearColor:false.
  // Its rect covers (50, 50), so a forced color clear there would wipe A's pixel.
  const viewB = new MapView({id: 'B', x: 40, width: 60, clear: true, clearColor: false});

  const viewState = {longitude: 0, latitude: 0, zoom: 1};
  const viewportA = viewA.makeViewport({width: 100, height: 100, viewState});
  const viewportB = viewB.makeViewport({width: 100, height: 100, viewState});

  const layer = new ScatterplotLayer({
    data: [{position: [0, 0]}],
    getPosition: d => d.position,
    radiusMinPixels: 10,
    pickable: true
  });

  const layerManager = new LayerManager(device, {viewport: viewportA});
  const pickLayersPass = new PickLayersPass(device);

  layerManager.setLayers([layer]);
  const {decodePickingColor} = pickLayersPass.render({
    viewports: [viewportA, viewportB],
    views: {A: viewA, B: viewB},
    // Render the object in view A only; view B just performs its clear.
    layerFilter: ({viewport}) => viewport.id === 'A',
    layers: layerManager.getLayers(),
    onViewportActive: layerManager.activateViewport,
    pickingFBO,
    deviceRect: {x: 0, y: 0, width: 100, height: 100}
  });

  const pixels = device.readPixelsToArrayWebGL(pickingFBO, {
    sourceX: 0,
    sourceY: 0,
    sourceWidth: 100,
    sourceHeight: 100
  });

  // (50, 50) is inside view B's rect (x >= 40). With clearColor:false the color must be
  // preserved, so view A's object is still picked there instead of being cleared away.
  const centerIndex = (50 * 100 + 50) * 4;
  const centerColor = pixels.slice(centerIndex, centerIndex + 4);
  const pickedObject = decodePickingColor!(centerColor);
  expect(
    pickedObject?.pickedLayer?.id,
    "bottom view's object survives the top view's clearColor:false clear"
  ).toBe(layer.id);
});
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by chrisgervang — COMMENTED** [review]
Hey, thanks for taking the time to fix this issue and adding a test to prevent it from recurring.

The fix looks well scoped. While looking for edge case, Claude found one I think is worth fixing to preserve semantics, and I had it add a comment for you to review.

**Comment by jarnoharno** [comment]
> Hey, thanks for taking the time to fix this issue and adding a test to prevent it from recurring.
> 
> The fix looks well scoped. While looking for edge case, Claude found one I think is worth fixing to preserve semantics, and I had it add a comment for you to review.

Thanks for the review. Good catch on the `clearColor: false case`. I applied the suggestion with a small simplification, letting the initial transparent `colorToUse` pass through when picking. I also added the suggested test. It doesn't fail on master, but it guards against an incorrect implementation. Both tests pass headless, and I confirmed the new test still fails against the previous unconditional version.

**Comment by chrisgervang** [comment]
Thanks! LGTM. The non-picking path should be equivalent after this change. Here's what how I expect the `clearColor` prop to behave after this lands:

clearColor | picking | result | correct?
-- | -- | -- | --
array | no | normalized color | ✓ (matches master)
array | yes | [0,0,0,0] (falls through) | ✓ no corruption
false | no | false (no color clear) | ✓
false | yes | false (no color clear) | ✓ semantics preserved
undefined | either | [0,0,0,0] | ✓

**Review by chrisgervang — APPROVED** [review]
